File: PHP/src/day01.php
<?php

namespace sjb\aoc2024;

require 'lib/Util.php';
// require 'lib/Geometry.php';
// require 'lib/Grid.php';

echo "Advent of Code 2024, Day 01: Historian Hysteria\n";

$INPUT_DIR = '../Input/';
// $INPUT_FILE = 'day01_test.txt';
$INPUT_FILE = 'day01_challenge.txt';
$INPUT_INDEX = 0;

$input = read_grouped_input($INPUT_DIR . $INPUT_FILE, $INPUT_INDEX);

$data = parse_data($input);

$result1 = solvePartOne($data);
echo "Part One: $result1\n";

$result2 = solvePartTwo($data);
echo "Part Two: $result2\n";


function solvePartOne(array $data):int {
	$distances = array_map(fn ($i): int => abs($data[0][$i] - $data[1][$i]), range(0, count($data[0])-1));
	$total = array_sum($distances);
	return $total;
}

function solvePartTwo(array $data):int  {
	$freq = array();
	foreach ($data[1] as $num) {
		if (!array_key_exists($num, $freq)) {
			$freq[$num] = 0;
		}
		$freq[$num] += 1;
	}
	$sim_score = 0;
	foreach ($data[0] as $num) {
		if (array_key_exists($num, $freq)) {
			$sim_score += $num * $freq[$num];
		}
	}
	return $sim_score;
}

function parse_data(array $input): array {
	$data = array();
	foreach ($input as $line) {
		$nums = preg_split('/\s+/', $line);
		array_push($data, $nums);
	}
	$data = pivot_matrix($data);
	
	sort($data[0], SORT_NUMERIC);
	sort($data[1], SORT_NUMERIC);
	
	return $data;
}

File: Perl/day01.pl
#!/usr/bin/env perl
use v5.40;
# use feature 'class';
# no warnings qw( experimental::class );

our $directory;
BEGIN { use Cwd; $directory = cwd; }
use lib $directory . '/lib';

use feature 'signatures';
use Data::Printer;
#use Storable 'dclone';

use AOC::Util;
#use AOC::Geometry;
#use AOC::Grid;

# my $INPUT_FILE = 'day01_test.txt';
my $INPUT_FILE = 'day01_challenge.txt';
my @input = read_input("../input/$INPUT_FILE");

say "Advent of Code 2024, Day 01: Historian Hysteria";

my @list1 = ();
my @list2 = ();

make_lists(@input);

solve_part_one();
solve_part_two();

exit( 0 );

sub solve_part_one(@input) {
	my $distance = 0;
	for my $i (0..$#list1) {
		my $d = abs($list2[$i] - $list1[$i]);
		$distance += $d;
	}
	say "Part One: The total distance is $distance";
}

sub solve_part_two() {
	# Turn List 2 into a frequency map
	my %map = {};
	for my $num (@list2) {
		$map{$num} ++;
	}

	my $similarity_score = 0;
	for my $num (@list1) {
		if (exists $map{$num}) {
			my $sim = $num * $map{$num};
			$similarity_score += $sim;
		}
	}
	say "Part Two: The similarity score is $similarity_score";
}

sub make_lists(@input) {
	for my $line (@input) {
		my @nums = split(/\s+/, $line);
		push(@list1, $nums[0]);
		push(@list2, $nums[1]);
	}
	@list1 = sort {$a <=> $b} @list1;
	@list2 = sort {$a <=> $b} @list2;
}